tools/smithy-rs-sync/src/main.rs

```diff
@@ -0,0 +1,409 @@
+fn set_last_synced_commit(repo: &Repository, oid: &Oid) -> Result<()> {
+    let repo_path = repo.workdir().expect("this will always exist");
+    let path = repo_path.join(COMMIT_HASH_FILENAME);
+    let mut file = OpenOptions::new().write(true).truncate(true).open(&path)?;
+
+    file.write(oid.to_string().as_bytes())
+        .with_context(|| format!("Couldn't write commit hash to '{}'", path.to_string_lossy()))?;
+
+    Ok(())
+}
+
+/// Run the necessary commands to build the SDK. On success, returns the path to the folder containing
+/// the build artifacts.
+fn build_sdk(smithy_rs_path: &Path) -> Result<PathBuf> {
+    println!("\tbuilding the SDK...");
+    let start = Instant::now();
+    let gradlew = smithy_rs_path.join("gradlew");
+
+    // The output of running this command isn't logged anywhere unless it fails
+    let clean_command_output = Command::new(&gradlew)
+        .arg("-Paws.fullsdk=true")
+        .arg(":aws:sdk:clean")
+        .current_dir(smithy_rs_path)
+        .output()
+        .with_context(|| {
+            format!(
+                "failed to execute '{} -Paws.fullsdk=true :aws:sdk:clean'",
+                gradlew.to_string_lossy()
+            )
+        })?;
+
+    if !clean_command_output.status.success() {
+        let stderr = String::from_utf8_lossy(&clean_command_output.stderr);
+        let stdout = String::from_utf8_lossy(&clean_command_output.stdout);
+
+        println!("stdout:\n{}\n", stdout);
+        println!("stderr:\n{}\n", stderr);
+
+        bail!("command to clean previous build artifacts from smithy-rs before assembling the SDK returned an error")
+    }
+
+    let assemble_command_output = Command::new(&gradlew)
+        .arg("-Paws.fullsdk=true")
+        .arg(":aws:sdk:assemble")
+        .current_dir(smithy_rs_path)
+        .output()
+        .with_context(|| {
+            format!(
+                "failed to execute '{} -Paws.fullsdk=true :aws:sdk:assemble'",
+                gradlew.to_string_lossy()
+            )
+        })?;
+
+    if !assemble_command_output.status.success() {
+        let stderr = String::from_utf8_lossy(&clean_command_output.stderr);
+        let stdout = String::from_utf8_lossy(&clean_command_output.stdout);
+
+        println!("stdout:\n{}\n", stdout);
+        println!("stderr:\n{}\n", stderr);
+
+        bail!("command to assemble the SDK returned an error")
+    }
+
+    let build_artifact_path = smithy_rs_path.join("aws/sdk/build/aws-sdk");
+
+    println!("\tsuccessfully built the SDK in {:?}", start.elapsed());
+    Ok(build_artifact_path)
+}
+
+/// Delete any current SDK files in aws-sdk-rust. Run this before copying over new files.
+fn clean_out_existing_sdk(aws_sdk_path: &Path) -> Result<()> {
+    println!("\tcleaning out previously built SDK...");
+
+    let sdk_path = format!("{}/sdk/*", aws_sdk_path.to_string_lossy());
+    let remove_sdk_command_output = Command::new("rm")
```

Comment:
It's a bit verbose to do it with stdlib, but instead of shelling out to copy and delete recursively, this crate is nice https://github.com/webdesus/fs_extra

Reply:
This crate is basically a Bash script, I'm going to leave it as is for simplicity's sake